src/liger_kernel/ops/fused_linear_jsd.py

```diff
@@ -0,0 +1,201 @@
+
+    grad_weight = (
+        torch.zeros_like(student_weight, device=device)
+        if student_weight.requires_grad
+        else None
+    )
+    grad_input = torch.zeros_like(student_input)
+    # we use fp32 for loss accumulator
+    loss_1d = torch.zeros((BT, V), dtype=torch.float32, device=device)
+
+    for chunk_id in range(num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = min((chunk_id + 1) * chunk_size, BT)
+
+        # chunk both inputs, shape: chunk_size x H
+        student_input_chunk = student_input[start_idx:end_idx]
+        teacher_input_chunk = teacher_input[start_idx:end_idx]
+
+        # when doing matmul, use the original precision, shape: chunk_size x V
+        student_logits_chunk = student_input_chunk @ student_weight.t()
+        teacher_logits_chunk = teacher_input_chunk @ teacher_weight.t()
+        chunk_n_rows = student_logits_chunk.shape[0]
+
+        # unreduced loss
+        loss_1d_slice = loss_1d[start_idx:end_idx]  # chunk_size
+        # log-softmax with temperature
+        student_logits_chunk = student_logits_chunk / temperature
+        teacher_logits_chunk = teacher_logits_chunk / temperature
+        student_prob_chunk = torch.log_softmax(student_logits_chunk, dim=-1)
+        teacher_prob_chunk = torch.log_softmax(teacher_logits_chunk, dim=-1)
+
+        # ensure _input and target are contiguous
+        student_prob_chunk = student_prob_chunk.contiguous()
+        teacher_prob_chunk = teacher_prob_chunk.contiguous()
+
+        # Here we calculate the gradient of prob_chunk in place so we can save memory.
+        _jsd_kernel[(chunk_n_rows,)](
+            X_ptr=student_prob_chunk,
+            X_stride=student_prob_chunk.stride(-2),
+            Y_ptr=teacher_prob_chunk,
+            Y_stride=teacher_prob_chunk.stride(-2),
+            loss_ptr=loss_1d_slice,
+            loss_stride=loss_1d_slice.stride(-2),
+            dX_ptr=student_prob_chunk,
+            dX_stride=student_prob_chunk.stride(-2),
+            beta=jsd_beta,
+            n_rows=BT,  # batchmean
+            n_cols=V,
+            BLOCK_SIZE=BLOCK_SIZE,
+        )
+        loss_1d[start_idx:end_idx] = loss_1d_slice
+        # gradients of prob_chunk in place, shape: chunk_size x V
+        # gradients of logits_chunk in place, shape: chunk_size x V
+        student_logits_chunk = (
+            student_prob_chunk
+            - torch.softmax(student_logits_chunk, dim=-1)
+            * student_prob_chunk.sum(dim=-1, keepdim=True).broadcast_to(
+                student_prob_chunk.shape
+            )
+        ) / temperature
+        grad_input[start_idx:end_idx] = student_logits_chunk @ student_weight
+
+        if grad_weight is not None:
+            torch.addmm(
+                input=grad_weight,
+                mat1=student_logits_chunk.t(),  # gradients of logits_chunk
+                mat2=student_input_chunk,
+                out=grad_weight,
+            )
+
+    loss = torch.sum(loss_1d) / BT
```

Comment:
I just noticed that torch.sum might overflow when BT size is large. It can be fixed by either
1. do division first before sum, i.e. torch.sum(loss_1d / BT) 
2. put division inside JSD kernel
I prefre the second solution more, but I'm not sure if it is ok to modify another kernel in this PR.

Reply:
can the [existing division in JSD](https://github.com/linkedin/Liger-Kernel/blob/de12602d858a6e83aaacc56e5cb64ab218c75a0a/src/liger_kernel/ops/jsd.py#L77) be able to handle this? 🤔 I saw you change the [`n_row` for each chunk ](https://github.com/linkedin/Liger-Kernel/pull/300/files#diff-1da1c5423f58c2cb144e1e7d131f17bc0eaf4915d34ced8177792dfee9f6aa46R85) to be BT now.

Reply:
curretnly n_row parameter in JSD is only for gradients calculation.
it can be used to loss calculation wrt demanding reduction method like cross_entropy
https://github.com/linkedin/Liger-Kernel/blob/de12602d858a6e83aaacc56e5cb64ab218c75a0a/src/liger_kernel/ops/cross_entropy.py#L143



> I saw you change the [`n_row` for each chunk ](https://github.com/linkedin/Liger-Kernel/pull/300/files#diff-1da1c5423f58c2cb144e1e7d131f17bc0eaf4915d34ced8177792dfee9f6aa46R85) to be BT now.

simply passing BT can perform correct calculations without further alpha tweaking like flce does
https://github.com/linkedin/Liger-Kernel/blob/de12602d858a6e83aaacc56e5cb64ab218c75a0a/src/liger_kernel/ops/fused_linear_cross_entropy.py#L112

Reply:
> can the [existing division in JSD](https://github.com/linkedin/Liger-Kernel/blob/de12602d858a6e83aaacc56e5cb64ab218c75a0a/src/liger_kernel/ops/jsd.py#L77) be able to handle this? 

it has potential overflow issue too, thats one of the reason why i think moving division into kernel is better

Reply:
Gotcha. I'm good with moving it inside the JSD kernel. For ce/flce, let's keep them for now. Thanks!